pkg/cmd/diff/diff.go

```diff
@@ -0,0 +1,159 @@
+package diff
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/MaineK00n/vuls-data-update/pkg/cmd/util/flag"
+	"github.com/MaineK00n/vuls-data-update/pkg/diff"
+	"github.com/MaineK00n/vuls-data-update/pkg/extract/util"
+	"github.com/MakeNowJust/heredoc"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+type option struct {
+	org       string
+	dir       string
+	filter    string
+	diffAlg   string
+	format    flag.Choices
+	outputDir string
+}
+
+func NewCmdDiff() *cobra.Command {
+	opts := &option{
+		org:     "vulsio",
+		dir:     filepath.Join(util.CacheDir(), "diff"),
+		filter:  "tree:0",
+		diffAlg: "default",
+		format:  flag.NewChoices([]string{"json", "text"}, "json"),
+	}
+
+	cmd := &cobra.Command{
+		Use:   "diff <datasource> <root ID> <old commit hash> <new commit hash>",
+		Short: "Show diff information",
+		Example: heredoc.Doc(`
+			$ vuls-data-update diff amazon ALAS2023-2023-059 95d80b3 c24e5ea
+			$ vuls-data-update diff nvd-api-cve CVE-2024-33892 e249e6c 0902d12
+		`),
+		Args: cobra.ExactArgs(4),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := run(cmd, args[0], args[1], args[2], args[3], opts); err != nil {
+				return errors.Wrap(err, "failed to diff")
+			}
+			return nil
+		},
+	}
+
+	cmd.Flags().StringVarP(&opts.dir, "dir", "d", filepath.Join(util.CacheDir(), "diff"), "working root directory in diff operation")
+	cmd.Flags().VarP(&opts.format, "format", "f", fmt.Sprintf("output format, one of [%s]", strings.Join(opts.format.Choices, ", ")))
+	cmd.Flags().StringVarP(&opts.outputDir, "output-dir", "o", "", "output to a file under specified dir. The file name is <rootID>-<kind>-<old>-<new>.(json|txt)")
+	cmd.Flags().StringVar(&opts.filter, "filter", "tree:0", "fiter for git clone and git fetch (e.g. tree:0, blob:none)")
+	cmd.Flags().StringVar(&opts.diffAlg, "diff-algorithm", "default", "algorithm for git diff c.f. https://git-scm.com/docs/git-diff")
+	cmd.Flags().StringVar(&opts.org, "org", "vulsio", "GitHub organization of repository")
+
+	_ = cmd.RegisterFlagCompletionFunc("format", opts.format.Completion)
+
+	// Just for completion, but don't validate the given value for diff-algorithm
+	// in case of git will add new algorithm options for diff.
+	_ = cmd.RegisterFlagCompletionFunc("diff-algorithm", func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		return []string{"default", "patience", "minimal", "histogram", "myers"}, cobra.ShellCompDirectiveDefault
+	})
+
+	return cmd
+}
+
+func run(cmd *cobra.Command, datasource, rootID, old, new string, opts *option) error {
+	// human friendly normalization
+	rootID = strings.TrimSuffix(filepath.Base(rootID), ".json")
+	datasource = strings.TrimPrefix(datasource, "vuls-data-extracted-")
```

Comment:
Is this necessary?
I don't think it's likely, but it appears to affect cases where the rootID ends in .json or the datasource ID starts with vuls-data-extracted-.

Reply:
Deleted, 008dc9c